lib/views/wallet/coin_details/faucet/faucet_button.dart

```diff
@@ -3,32 +3,104 @@ import 'package:flutter/material.dart';
 import 'package:web_dex/common/screen.dart';
 import 'package:web_dex/generated/codegen_loader.g.dart';
 import 'package:web_dex/shared/ui/ui_primary_button.dart';
+import 'package:komodo_defi_types/komodo_defi_types.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:web_dex/bloc/faucet_button/faucet_button_bloc.dart';
+import 'package:web_dex/bloc/faucet_button/faucet_button_event.dart';
+import 'package:web_dex/bloc/faucet_button/faucet_button_state.dart';
+import 'package:web_dex/views/wallet/coin_details/faucet/faucet_view.dart';
 
-class FaucetButton extends StatelessWidget {
+class FaucetButton extends StatefulWidget {
   const FaucetButton({
-    Key? key,
-    required this.onPressed,
+    super.key,
+    required this.coinAbbr,
+    required this.address,
     this.enabled = true,
-  }) : super(key: key);
+  });
 
   final bool enabled;
-  final VoidCallback onPressed;
+  final String coinAbbr;
+  final PubkeyInfo address;
 
+  @override
+  State<FaucetButton> createState() => _FaucetButtonState();
+}
+
+class _FaucetButtonState extends State<FaucetButton> {
   @override
   Widget build(BuildContext context) {
     final ThemeData themeData = Theme.of(context);
-    return UiPrimaryButton(
-      key: const Key('coin-details-faucet-button'),
-      height: isMobile ? 52 : 40,
-      textStyle: themeData.textTheme.labelLarge
-          ?.copyWith(fontSize: 14, fontWeight: FontWeight.w600),
-      backgroundColor: themeData.colorScheme.tertiary,
-      text: LocaleKeys.faucet.tr(),
-      prefix: const Padding(
-        padding: EdgeInsets.only(right: 10),
-        child: Icon(Icons.local_drink_rounded, color: Colors.blue),
-      ),
-      onPressed: !enabled ? null : onPressed,
+
+    return BlocConsumer<FaucetBloc, FaucetState>(
+      listenWhen: (previous, current) {
+        return (current is FaucetLoading && current.address == widget.address.address) ||
```

Comment:
Nit: I'd recommend refactoring out the `listenWhen` conditions into variables to make the condition simpler to understand at first glance.

For example:
```suggestion
        final isLoading = current is FaucetLoading && current.address == widget.address.address;
        final didStopLoading = previous is FaucetLoading && current is! FaucetLoading;
        
        return isLoading || didStopLoading;
```